### ISSUE TITLE
'YDocWebSocketHandler' object has no attribute 'room'

### ISSUE THREAD

**Issue by martinRenou** [jupyterlab/jupyter-collaboration#321]
I'm getting the following error with jupyter-collaboration 2.1.1:

```
    Traceback (most recent call last):
      File "/home/martin/micromamba/envs/jupytercad/lib/python3.12/site-packages/tornado/ioloop.py", line 750, in _run_callback
        ret = callback()
              ^^^^^^^^^^
      File "/home/martin/micromamba/envs/jupytercad/lib/python3.12/site-packages/tornado/websocket.py", line 640, in <lambda>
        self.stream.io_loop.add_future(result, lambda f: f.result())
                                                         ^^^^^^^^^^
      File "/home/martin/micromamba/envs/jupytercad/lib/python3.12/site-packages/jupyter_collaboration/handlers.py", line 286, in on_message
        changes = self.room.awareness.get_changes(message[1:])
                  ^^^^^^^^^
    AttributeError: 'YDocWebSocketHandler' object has no attribute 'room'
```

**Comment by davidbrochart** [comment]
Could you try updating to jupyter-collaboration v3.0.0.a2?

**Comment by martinRenou** [comment]
When using jupyter-collaboration 3.0.0a2 I'm getting a front-end error when trying to open a file 

![Screenshot from 2024-07-02 11-37-29](https://github.com/jupyterlab/jupyter-collaboration/assets/21197331/af265e30-4726-45ab-aea2-1692942279bb)

**Comment by krassowski** [comment]
Is this when opening a plain text file or when opening the CAD file?

**Comment by martinRenou** [comment]
It was when opening a CAD file. We should try to upgrade jupyter collaboration there with the recent beta.

**Comment by krassowski** [comment]
Note that the content (`"test")` is not a valid notebook. If I replace it with

```
nb = nbformat.v4.new_notebook()
content = nbformat.writes(nb, version=4)
```

then it does not throw. It looks like exception handling may be off here.

**Comment by fperez** [comment]
Any progress on this? On a clean, fresh install of jupyterlab with python 3.12 with mambaforge, I basically can't open _any_ existing notebooks as I get pages and pages of these errors. I had to uninstall the collab machinery entirely so I could at least get work done.

Given this is on a fully clean install on an otherwise empty system, I'm pretty concerned that it is likely to hit users very easily... And it's a hard blocker - nothing works, documents can't be opened, nor even closed: the UI hangs in a weird state.

**Comment by krassowski** [comment]
As per my previous reply, I think this error message is a read herring. It is basically error which shows up when the connection closes if there is a different error which lead to the connection close. The immediate action item is to expose the real error so that we can address potentially multiple underlying problems. Do you also see this on newly created notebooks or only on existing notebooks? If it is the latter, can you share such an existing notebook?

I do not experience the problem you are facing but none of the deployments I manage use Python 3.12, and I am using the bleeding edge beta of version 3. I hope that with version 3/JupyterLab 4.3 - also in beta - many bugs will be ironed out to make it production-ready (for now I would not advise trying it yet as there are known issues in works).

**Comment by martinRenou** [comment]
From using more RTC features, I seem to be able to get a clean working setup when starting fresh and removing the `.jupyter_ystore.db` file at the location of the server. @fperez you may want to try removing that file and restarting the server.

**Comment by fperez** [comment]
Mmh, on a completely clean install from scratch, made right now by downloading miniforge3 and adding [my default packages](https://github.com/fperez/homesetup/blob/main/usr/bin/setup/environment.yml), and **after deleting the `jupyter_ystore.db` file** at the root of the server, I get a completely non-functional Jupyter. It fails to open *any* existing notebook or to create any new one, always with the same error:

![image](https://github.com/user-attachments/assets/95fb7d61-691d-4c32-aced-959ad84c389b)

Basically (at least on this machine), the only way to use Jupyter for anything is to entirely remove `jupyter-collaboration` 2.1.0...  I hope it's just a glitch on my machine (macOS 15.0.1, Firefox, all up to date), but if it's reproducible, it's pretty bad.

**Comment by davidbrochart** [comment]
@fperez I created an environment using your [environment.yml](https://raw.githubusercontent.com/fperez/homesetup/9cd02d01851b97cc516dd1e3b8fcbdc2a81f4833/usr/bin/setup/environment.yml) using `pixi`, and everything seems to work fine:
```bash
pixi init --import environment.yml
pixi shell
jupyter lab
```
It's strange that you are mentioning `jupyter-collaboration` v2.1.0, because it should have installed v2.1.4. Could you try again?

**Comment by fperez** [comment]
Ah, you may be onto something... It seems the culprit is a version pin in JupyterLab:

```
The following packages are incompatible
├─ jupyter-collaboration is installable with the potential options
│  ├─ jupyter-collaboration 2.1.4 would require
│  │  └─ jupyterlab >=4.0.5,<4.2 , which can be installed;
│  └─ jupyter-collaboration [1.0.0|1.0.1|...|2.1.3] conflicts with any installable versions previously reported;
└─ jupyterlab 4.2.5**  is not installable because it conflicts with any installable versions previously reported.
```

I had Lab 4.2.5, which kept me pinned at collab 2.1.0. I can't upgrade jupyter-collaboration to 2.1.4 without downgrading JupyterLab.

**Comment by fperez** [comment]
I can confirm that once I allow the JuptyerLab downgrade to 4.1.8 and install collab 2.1.4, it works fine.

So whatever the bug was in collab was indeed fixed, thanks! The problem we have now is this version pin in JupyterLab, which may end up leaving users locked into an older version of collab with the bug.

Thanks for the pointer!

**Comment by davidbrochart** [comment]
It seems to be a conda issue, because with `pip` one can install `jupyterlab==4.2.5` and `jupyter-collaboration==2.1.4` in the same environment.

**Comment by davidbrochart** [comment]
I opened https://github.com/conda-forge/jupyter-collaboration-feedstock/pull/16.

**Comment by davidbrochart** [comment]
Actually conda-forge was doing the right thing. I released jupyter-collaboration v2.1.5 which fixes the issue.

**Comment by fperez** [comment]
Awesome, thanks so much @davidbrochart!! Glad we tracked this down :)

**Comment by davidbrochart** [comment]
Thank you for reporting the issue!
